### PR TITLE
Take over screen for usernames

### PR DESCRIPTION
**Pull request by JPonte**
#### APK
[Download build #8017](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8017/artifact/build/artifact/wire-dev-PR364-8017.apk)
[Download build #8018](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8018/artifact/build/artifact/wire-dev-PR364-8018.apk)
[Download build #8019](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8019/artifact/build/artifact/wire-dev-PR364-8019.apk)
[Download build #8021](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8021/artifact/build/artifact/wire-dev-PR364-8021.apk)
[Download build #8023](http://192.168.10.18:8080/job/Pull%20Request%20Builder/8023/artifact/build/artifact/wire-dev-PR364-8023.apk)

### CODE REVIEW

File: app/src/androidTest/java/com/waz/zclient/startup/fragments/UsernamesTakeOverScreenTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        pressBack();
+        Thread.sleep(500);
+        verify(activity.getFragmentManager().findFragmentByTag(FirstTimeAssignUsername.TAG) != null);
+    }
+
+    @Test
+    public void verifyConversationListIsOpenedAfterAcceptingUsername() throws InterruptedException {
+        attachFragment(FirstTimeAssignUsername.newInstance("", ""), FirstTimeAssignUsername.TAG);
+        Thread.sleep(500);
+        onView(withId(R.id.zb__username_first_assign__keep)).perform(click());
```
Comment (shumengye) [inline app/src/androidTest/java/com/waz/zclient/startup/fragments/UsernamesTakeOverScreenTest.java]: how does the test verify that the conversation list is shown?

File: app/src/main/java/com/waz/zclient/MainActivity.java [hunk 2]
```diff
@@ -493,6 +493,9 @@ public void run() {
             }
             IntentUtils.clearLaunchIntentExtra(intent);
         }
+
+        getControllerFactory().getUsernameController().setUser(self);
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/MainActivity.java]: Move this to beginning of method where other stores are set too

File: app/src/androidTest/java/com/waz/zclient/startup/fragments/UsernamesTakeOverScreenTest.java (hunk 1 above)
Comment (JPonte) [inline app/src/androidTest/java/com/waz/zclient/startup/fragments/UsernamesTakeOverScreenTest.java]: True, there should be a verify() that it closed there.

File: app/src/main/java/com/waz/zclient/MainActivity.java [hunk 3]
```diff
@@ -741,6 +744,7 @@ public void logout() {
         getControllerFactory().getTrackingController().tagEvent(new SignOut());
         getControllerFactory().getTrackingController().tagEvent(new LoggedOutEvent());
         getStoreFactory().getZMessagingApiStore().logout();
+        getControllerFactory().getUsernameController().logout();
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/MainActivity.java]: Add this also to the logout method in `AccountPreferences`

File: app/src/main/java/com/waz/zclient/controllers/usernames/UsernamesController.java [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Random;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+public class UsernamesController implements IUsernamesController {
+
+    private static final int MAX_ATTEMPTS = 10;
```
Comment (shumengye) [inline app/src/main/java/com/waz/zclient/controllers/usernames/UsernamesController.java]: Is 10 attempts enough for the thousands of "Maria"s?

Comment (JPonte) [inline app/src/main/java/com/waz/zclient/controllers/usernames/UsernamesController.java]: Maybe... But I can raise it. The random number has 4 digits so 10000 different possibilities. Not sure how likely is it that the 10 you get will be already used.

**Comment by JPonte** [comment]
build again

**Pull request merged**